### PR TITLE
Salesforce objects

### PR DESCRIPTION
**Pull request by mqueme**
Please answer the following questions:

| Q             | A
| ------------- | ---
| Bug fix?      | n
| New feature?  | y
| BC breaks?    | n
| Deprecations? | n
| Fixed issues  |  n
| User Documentation| https://github.com/mautic/documentation/pull/76

## Description
This PR, add functionality to Salesforce plugin.  The user will now be able to select what salesforce objects to pull data from and import them to Mautic contacts.

## Steps to test this PR
Apply this PR, in the salesforce configuration item you should now see the objects Lead and Contact.

The Salesforce CLI command will now fetch data from the selected objects and import them into Mautic Contacts. Pushing data will still continue to go into Salesforce Leads.

you need a secure site with domain name to test this: I have added a self signed certificate to my localhost and changed the name to a qualified domain name.

To test pulling leads run this command on command line:

php app/console mautic:integration:fetchleads --time-interval="29 days" --integration=Salesforce

This should import leads and/or contacts to mautic contacts

To test pushing lead activities
First create the salesforce object. you can follow steps here: https://github.com/mautic/documentation/pull/76/files

then run this command
php app/console mautic:integration:pushleadactivity --time-interval="29 days" --integration=Salesforce

this should push points, email activity, and form activity to the mautic object on salesforce to the leads that we have synced
make sure you update your namespace for the mautic object created in salesforce

### CODE REVIEW

File: app/bundles/PointBundle/Entity/PointRepository.php [hunk 1]
```diff
@@ -98,7 +98,35 @@ public function getCompletedLeadActions($type, $leadId)
 
         return $return;
     }
+    /**
+     * @param int    $leadId
+     *
+     * @return array
+     */
+    public function getCompletedLeadActionsByLeadId($leadId)
```
Comment (alanhartless) [inline app/bundles/PointBundle/Entity/PointRepository.php]: Is this method used anywhere?

File: app/bundles/PluginBundle/Command/PushLeadActivityCommand.php [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+        if ($integration && $startDate && $endDate) {
+            /** @var \Mautic\PluginBundle\Helper\IntegrationHelper $integrationHelper */
+            $integrationHelper = $factory->getHelper('integration');
+
+            $integrationObject = $integrationHelper->getIntegrationObject($integration);
+
+            if ($integrationObject !== null && method_exists($integrationObject, 'pushLeadActivity')) {
+
+                $output->writeln('<info>'.$translator->trans('mautic.plugin.command.fetch.leads', array('%integration%' => $integration)).'</info>');
```
Comment (alanhartless) [inline app/bundles/PluginBundle/Command/PushLeadActivityCommand.php]: Should this be a different language string than "fetch?" Same in line 105

File: app/bundles/PluginBundle/Command/PushLeadActivityCommand.php [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+            ->addOption('--start-date', '-d', InputOption::VALUE_REQUIRED, 'Set start date for updated values.')
+            ->addOption(
+                '--end-date',
+                '-t',
+                InputOption::VALUE_REQUIRED,
+                'Set end date for updated values.'
+            )
+            ->addOption(
+                '--time-interval',
+                '-ti',
```
Comment (alanhartless) [inline app/bundles/PluginBundle/Command/PushLeadActivityCommand.php]: Symfony won't like this as it'll throw a "Too many arguments" error. It has to be single letter that's not used.

File: plugins/MauticCrmBundle/Integration/SalesforceIntegration.php [hunk 4]
```diff
@@ -202,25 +251,365 @@ public function getFetchQuery($params){
...
     {
-        $fields = array_keys($this->getAvailableLeadFields());
+        $settings['feature_settings']['objects']=$object;
+        $fields = array_keys($this->getAvailableLeadFields($settings));
+
         $params['fields']=implode(',',$fields);
 
+        $fields = array_keys($params['fields']);
+
+        $leadFields = $this->cleanSalesForceData($params,$fields, $object);
```
Comment (alanhartless) [inline plugins/MauticCrmBundle/Integration/SalesforceIntegration.php]: Doesn't look like this is used anywhere?

File: app/migrations/Version20160808000000.php [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+/**
+ * Auto-generated Migration: Please modify to your needs!
+ */
+class Version20160806000000 extends AbstractMigration
+{
+    /**
+     * @param Schema $schema
+     */
+    public function up(Schema $schema)
```
Comment (alanhartless) [inline app/migrations/Version20160808000000.php]: This should be `preUp`

File: app/migrations/Version20160808000000.php [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        if ($schema->getTable($this->prefix.'integration_entity')) {
+
+            throw new SkipMigrationException('Schema includes this migration');
+        }
+    }
+
+    /**
+     * @param Schema $schema
+     */
+    public function down(Schema $schema)
```
Comment (alanhartless) [inline app/migrations/Version20160808000000.php]: This should be `up`

File: app/bundles/PluginBundle/Entity/IntegrationEntity.php [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+    /**
+     * @param ORM\ClassMetadata $metadata
+     */
+    public static function loadMetadata (ORM\ClassMetadata $metadata)
+    {
+        $builder = new ClassMetadataBuilder($metadata);
+
+        $builder->setTable('integration_entity')
+            ->setCustomRepositoryClass('Mautic\PluginBundle\Entity\IntegrationEntityRepository');
+
```
Comment (alanhartless) [inline app/bundles/PluginBundle/Entity/IntegrationEntity.php]: Let's add some indexes on this table: [integration, integration_entity, internal_entity] and [integration, integration_entity, integration_entity_id] (don't forget the migration)

File: app/bundles/PluginBundle/Entity/IntegrationEntity.php [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+        $builder->addId();
+
+        $builder->addDateAdded();
+
+        $builder->addNullableField('integration', 'array');
+
+        $builder->createField('integrationEntity', 'string')
+            ->columnName('inegration_entity')
+            ->build();
+        $builder->createField('integrationEntity', 'integer')
```
Comment (alanhartless) [inline app/bundles/PluginBundle/Entity/IntegrationEntity.php]: I think this was meant to be `integrationEntityId`

File: app/bundles/PluginBundle/Entity/IntegrationEntity.php [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+    /**
+     * @param mixed $isMissing
+     */
+    public function setInternalEntityId ($internalEntityId)
+    {
+        $this->internalEntityId = $internalEntityId;
+    }
+
+    /**
+     * @return mixed
```
Comment (alanhartless) [inline app/bundles/PluginBundle/Entity/IntegrationEntity.php]: Is setInternal() and getInternal() used? They aren't defined loadMetadata() or the migration so just validating.

File: app/bundles/PluginBundle/Entity/IntegrationEntityRepository.php [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+use Mautic\CoreBundle\Entity\CommonRepository;
+
+/**
+ * IntegrationRepository
+ */
+class IntegrationEntityRepository extends CommonRepository
+{
+
+    public function getIntegrationsEntityId($integration, $integrationEntity, $internalEntity, $internalEntityId)
```
Comment (alanhartless) [inline app/bundles/PluginBundle/Entity/IntegrationEntityRepository.php]: `$internalEntityId` is required but there is at least one use of getIntegrationsEntityId() that isn't passing in a fourth argument. Was `$internalEntityId` meant to be defaulted to null?

File: app/bundles/CoreBundle/Entity/AuditLogRepository.php [hunk 11]
```diff
@@ -65,6 +65,23 @@ public function getLogForObject($object = null, $id = null, $limit = 10, $afterD
         return $query->getQuery()->getArrayResult();
     }
 
+    public function getDetailsByAction($object = null, $bundle = null, $action = null)
```
Comment (alanhartless) [inline app/bundles/CoreBundle/Entity/AuditLogRepository.php]: I think this can be removed now? Or is it used somewhere else?

**Comment by alanhartless** [comment]
+1

**Comment by escopecz** [comment]
+1

**Pull request merged**